### DISCUSSION TITLE
Add plpythonu3 extension

### DISCUSSION THREAD

**Discussion by Vinzent03** [supabase/supabase#2239] [Questions] [answered]
I am **really** new to sql in general and want to use python in my functions, because I need to use postgis. It's not listed in the extensions list and enabling via 

`CREATE EXTENSION plpythonu3` throws the following error: 

`could not open extension control file "/usr/share/postgresql/12/extension/plpythonu3.control": No such file or directory`

**Comment by inian** [comment, marked as answer]
Hi @Vinzent03, we are not planning to offer [untrusted extensions](https://www.postgresql.org/docs/9.1/plpython.html) in Supabase in the near future. It is hard to limit the functionality of these untrusted extensions and a malicious user can make the platform unsafe for everyone.

**Reply by aantn** [reply]
@inian plpython3u is listed on https://supabase.io/docs/guides/database/extensions. Is this a mistake?

**Reply by inian** [reply]
yup thats a mistake @aantn. Updating the docs now. You can find the most up to date list of extensions on your dashboard https://app.supabase.io/project/your-project-id/database/extensions

**Reply by bitnom** [reply]
If only superusers can create functions in untrusted languages, then how does it make the platform unsafe for everyone? It's sort of like saying you don't trust your customers to be responsible admins of their own data. There are currently many ways we can misconfigure our databases to be insecure, and that's on us. I for one would appreciate the ability to decide for myself. I know I can write a secure python function. Just warn the user in the description and in the docs. Maybe only allow it for pro/paying accounts.

**Reply by inian** [reply]
@bitnom, we do not give access to superuser directly in our hosted platform. It is very hard to restrict what a superuser can do within Postgres which is why all hosted postgres platforms limit the user to some extent. Superusers can execute commands, read arbitrary files within the database server which could potentially compromise the security of other databases hosted on Supabase. That is the primary reason why we block superuser access (which limits the usefulness of untrusted extensions).

**Comment by benjamin-kirkbride** [comment]
Why is the Python extension a problem but the JS and Java ones are not?

**Reply by GaryAustin1** [reply]
https://www.postgresql.org/docs/current/plpython.html   mentions it is untrusted...
`PL/Python is only available as an “untrusted” language, meaning it does not offer any way of restricting what users can do in it and is therefore named plpython3u. A trusted variant plpython might become available in the future if a secure execution mechanism is developed in Python. The writer of a function in untrusted PL/Python must take care that the function cannot be used to do anything unwanted, since it will be able to do anything that could be done by a user logged in as the database administrator. Only superusers can create functions in untrusted languages such as plpython3u.`
For plv8:
https://plv8.github.io/
`PLV8 is a trusted Javascript language extension for PostgreSQL. It can be used for stored procedures, triggers, etc.`

**Comment by omani** [comment]
for those interested in how to enable plpython3u **for self-hosted** supabase via docker.

remove the image from docker-compose.yml and extend it by using a custom Dockerfile using the same image.
note that this will also update the `postgresql-15` package but that shouldnt be a problem since supabase (afaik) builds *around* the vanilla postgresql package.

here is how to install plpython3u for supabase:

```bash
diff --git a/docker/docker-compose.yml b/docker/docker-compose.yml
index 5060991..b3041b6 100644
--- a/docker/docker-compose.yml
+++ b/docker/docker-compose.yml
@@ -378,7 +378,6 @@ services:
   # Comment out everything below this point if you are using an external Postgres database
   db:
     container_name: supabase-db
-    image: supabase/postgres:15.1.0.103
     healthcheck:
       test: pg_isready -U postgres -h localhost
       interval: 5s
@@ -410,6 +409,9 @@ services:
       POSTGRES_PASSWORD: ${POSTGRES_PASSWORD}
       PGDATABASE: ${POSTGRES_DB}
       POSTGRES_DB: ${POSTGRES_DB}
+    build:
+      context: .
+      dockerfile: "./postgres/Dockerfile"
     volumes:
       - ./volumes/db/realtime.sql:/docker-entrypoint-initdb.d/migrations/99-realtime.sql:Z
       # Must be superuser to create event trigger
```

```
$ cat postgres/Dockerfile
FROM supabase/postgres:15.1.0.103

RUN /usr/share/postgresql-common/pgdg/apt.postgresql.org.sh -y
RUN apt update
RUN apt install -y postgresql-plpython3-15
```

```
$ docker compose build
$ docker compose up
```

Now you will find the extension listed on the dashboard under http://localhost:8000/project/default/database/extensions
